src/main/java/org/javarosa/core/util/externalizable/ExtWrapArrayListMultiMap.java

```diff
@@ -0,0 +1,93 @@
+package org.javarosa.core.util.externalizable;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+
+public class ExtWrapArrayListMultiMap extends ExternalizableWrapper {
+
+    private ExternalizableWrapper keyType;
+
+    /* Constructors for serialization */
+
+    public ExtWrapArrayListMultiMap(Multimap val) {
+        this(val, null);
+    }
+
+    public ExtWrapArrayListMultiMap(Multimap val, ExternalizableWrapper keyType) {
+        if (val == null) {
+            throw new NullPointerException();
+        }
+
+        this.val = val;
+        this.keyType = keyType;
+    }
+
+    /* Constructors for deserialization */
+
+    public ExtWrapArrayListMultiMap() {
+    }
+
+
+    public ExtWrapArrayListMultiMap(Class keyType) {
+        this(new ExtWrapBase(keyType));
+    }
+
+    public ExtWrapArrayListMultiMap(ExternalizableWrapper keyType) {
+        if (keyType == null) {
+            throw new NullPointerException();
+        }
+
+        this.keyType = keyType;
+    }
+
+    @Override
+    public ExternalizableWrapper clone(Object val) {
+        return new ExtWrapArrayListMultiMap((Multimap)val, keyType);
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
+        long size = ExtUtil.readNumeric(in);
+        ArrayListMultimap<Object, Object> multimap = ArrayListMultimap.create();
+        for (int i = 0; i < size; i++) {
+            Object key = ExtUtil.read(in, keyType, pf);
+
+            long numberOfValues = ExtUtil.readNumeric(in);
+            for (long l = 0; l < numberOfValues; l++) {
+                multimap.put(key, ExtUtil.read(in, new ExtWrapTagged(), pf));
+            }
+        }
+        val = multimap;
+    }
+
+    @Override
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ArrayListMultimap multimap = (ArrayListMultimap)val;
```

Comment:
It seems a bit unnecessary to me that this whole wrapper be dependent on the implementation of the multimap, and potentially very dangerous for the future if code signatures which use just the multimap signature change their underlying implementation.

I'd suggest doing a type cast here to an arraylist multimap by checking whether that's the implementation first and casting if so, but converting the object type if not, rather than crashing.